src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -184,15 +323,137 @@ static P11Key convertKey(Token token, Key key, String algo,
         if (p11Key != null) {
             return p11Key;
         }
-        if (!"RAW".equalsIgnoreCase(key.getFormat())) {
-            throw new InvalidKeyException("Encoded format must be RAW");
+        if (key instanceof PBEKey pbeKey) {
+            ki = ki == null ? getKeyInfo(keyAlgo) : ki;
+            if (ki instanceof PBEKeyInfo pbeKi) {
+                try {
+                    p11Key = derivePBEKey(token, getPbeKeySpec(pbeKey), pbeKi);
+                } catch (InvalidKeySpecException e) {
+                    throw new InvalidKeyException(e);
+                }
+            } else {
+                throw new InvalidKeyException("Cannot derive unknown " +
+                        keyAlgo + " algorithm");
+            }
+        } else {
+            if (si instanceof PBEKeyInfo) {
+                throw new InvalidKeyException("PBE service requires a PBE key");
+            }
+            if (!"RAW".equalsIgnoreCase(key.getFormat())) {
+                throw new InvalidKeyException("Encoded format must be RAW");
+            }
+            byte[] encoded = key.getEncoded();
+            p11Key = createKey(token, encoded, svcAlgo, si.keyType, extraAttrs);
         }
-        byte[] encoded = key.getEncoded();
-        p11Key = createKey(token, encoded, algo, algoType, extraAttrs);
         token.secretCache.put(key, p11Key);
         return p11Key;
     }
 
+    static P11Key derivePBEKey(Token token, PBEKeySpec keySpec,
+            PBEKeyInfo pbeKi) throws InvalidKeySpecException {
+        token.ensureValid();
+        if (keySpec == null) {
+            throw new InvalidKeySpecException("PBEKeySpec must not be null");
+        }
+        Session session = null;
+        try {
+            session = token.getObjSession();
+            CK_MECHANISM ckMech;
+            char[] password = keySpec.getPassword();
```

Comment:
Good

Reply:
I've just noticed that in this case in particular we can clean it up here but we need to save a copy in P11PBEKey because if the key has to be transferred to a different P11 token, we need to re-derive from the password, salt and iteration count. This case would happen for example if you have a P11 key from one token and you want to use it in a P11 service from a different token. Code is in P11SecretKeyFactory::convertKey. For safety, I'll clone the password in the P11PBEKey constructor.

Reply:
Update: we finally decided to do what I described in my previous comment but in those PBE MAC and Cipher cases in which we know that the key was derived for the same token than the service AND the key cannot be accessed from outside —in other words, there are strong guarantees that the key won't need to be derived again—, we will clear the password.